File: pulse-kotlin/src/commonMain/kotlin/com/fluxchess/pulse/kotlin/engine/MoveList.kt
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package com.fluxchess.pulse.kotlin.engine

class MoveList {
	var size: Int = 0
	val entries: Array<MoveEntry> = Array(MAX_PLY) { MoveEntry() }

	internal fun reset() {
		size = 0
	}

	internal fun add(move: Move) {
		entries[size++].move = move
	}

	internal fun rateByMVVLVA() {
		for (i in 0 until size) {
			val move = entries[i].move
			var value = 0

			val pieceTypeValue = pieceTypeValueOf(pieceTypeOf(originPieceOf(move)))
			value += KING_VALUE / pieceTypeValue

			val targetPiece = targetPieceOf(move)
			if (targetPiece != NO_PIECE) {
				value += 10 * pieceTypeValueOf(pieceTypeOf(targetPiece))
			}

			entries[i].value = value
		}
	}

	internal fun sort() {
		for (i in 0 until size) {
			val entry = entries[i]

			var j = i
			while (j > 0 && entries[j - 1].value < entry.value) {
				entries[j] = entries[j - 1]
				j--
			}

			entries[j] = entry
		}
	}
}

class MoveEntry {
	var move: Move = NO_MOVE
	internal var value: Value = NO_VALUE
}

File: pulse-go/internal/pulse/engine/move_list.go
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package engine

type MoveList struct {
	Size    int
	Entries [MaxPly]MoveEntry
}

func (ml *MoveList) reset() {
	ml.Size = 0
}

func (ml *MoveList) add(m Move) {
	ml.Entries[ml.Size].Move = m
	ml.Size++
}

func (ml *MoveList) rateByMVVLVA() {
	for i := 0; i < ml.Size; i++ {
		m := ml.Entries[i].Move
		v := 0

		pieceTypeValue := pieceTypeValueOf(PieceTypeOf(OriginPieceOf(m)))
		v += kingValue / pieceTypeValue

		targetPc := TargetPieceOf(m)
		if targetPc != NoPiece {
			v += 10 * pieceTypeValueOf(PieceTypeOf(targetPc))
		}

		ml.Entries[i].value = v
	}
}

func (ml *MoveList) sort() {
	for i := 0; i < ml.Size; i++ {
		entry := ml.Entries[i]

		j := i
		for j > 0 && ml.Entries[j-1].value < entry.value {
			ml.Entries[j] = ml.Entries[j-1]
			j--
		}

		ml.Entries[j] = entry
	}
}

type MoveEntry struct {
	Move  Move
	value value
}
